---
output: html_document
editor_options: 
  chunk_output_type: console
---
```{r}
#install.packages("tidyverse")
library(tidyverse)
#install.packages("influential")
library(influential)
#install.packages("bigreadr")
library(bigreadr)
library(bhr)
```


```{r}
ac_variantlevel_asc <- bigreadr::fread2("ASC_variant_results.tsv.bgz")
ac_variantlevel_asc
baseline_model <- read.table("ms_baseline_oe5.txt")
# Convert the data to a tibble and apply filters
ac_variantlevel_asc <- ac_variantlevel_asc %>% 
  as_tibble() %>%
  filter(in_analysis == TRUE, !is.na(gene_id)) %>%
  filter(locus %>% str_detect("X|Y", negate = TRUE))

# subset dbs group
dbs_variantlevel_asc = ac_variantlevel_asc %>% 
  filter(group=="DBS") %>% 
  select(gene_id, consequence, ac_case, ac_ctrl, locus, mpc)


wrangle_sumstats <- function(table,n_cases,n_controls, var_filter) {
  
  #Filter to variants of interest
  table = table[var_filter,] 
  
  #Compute sample prevalence, will be used to compute per-sd beta
  prevalence = n_cases/(n_cases+n_controls) 
  
  #Compute variant MAF in cases, will be used to compute per-sd beta
  table$AF_case = table$ac_case/(2*n_cases) 
  
  #Compute variant MAFoverall, will be used to compute per-sd beta
  table$AF = (table$ac_case + table$ac_ctrl)/(2*(n_cases + n_controls)) 
  
  #calculate per-sd betas
  table$beta = (2 * (table$AF_case - table$AF) * prevalence)/sqrt(2 * table$AF * (1 - table$AF) * prevalence * (1 - prevalence))  
  
  #calculate variant variances
  table$twopq = 2*table$AF * (1 - table$AF) 
  
  #convert betas from per-sd (i.e. sqrt(variance explained)) to per-allele (i.e. in units of phenotype) betas.
  #per-allele are the usual betas reported by an exome wide association study.
  table$beta_perallele = table$beta/sqrt(table$twopq)
  
  #aggregate into gene-level table.
  #position doesn't need to be super precise, as it is only used to order genes for jackknife
  #N = sum of case and control counts
  sumstats = data.frame(gene = table$gene_id,
                        AF = table$AF,
                        beta = table$beta_perallele,
                        gene_position = parse_number(sapply(strsplit(table$locus, split = ":"), function(x) x[[2]])),
                        chromosome =  parse_number(sapply(strsplit(table$locus, split = ":"), function(x) x[[1]])),
                        N = n_cases + n_controls,
                        phenotype_key = "BP")
  
  #we have found that in these smaller sample analyses, there are some genes with
  #large burden scores that are clearly outliers
  #we remove one such gene, TTN
  sumstats <- sumstats[sumstats$gene != "ENSG00000155657",]
  
  
  return(sumstats)
}


# Analyze PTV and Missense for DBS 
n_cases <- 4811
n_controls <- 5214

dbs_variantlevel_asc %>% 
  count(consequence)

# PTV Analysis for DBS
dbs_sumstats_ptv <- wrangle_sumstats(dbs_variantlevel_asc, n_cases, n_controls,
                                     dbs_variantlevel_asc$consequence %in% c("frameshift_variant", "stop_gained", "splice_acceptor_variant", "splice_donor_variant", "start_lost","stop_lost"))
saveRDS(dbs_sumstats_ptv, file = "ac_sumstats_ptv.rds")
# Missense (MPC >= 2) Analysis for DBS
dbs_sumstats_missense <- wrangle_sumstats(dbs_variantlevel_asc, n_cases, n_controls,
                                          dbs_variantlevel_asc$consequence == "missense_variant" & !is.na(dbs_variantlevel_asc$mpc) & dbs_variantlevel_asc$mpc >= 2)


# Calculate heritability for each group and variant type using BHR
dbs_ptv_bhr <- BHR(trait1_sumstats = dbs_sumstats_ptv, annotations = list(baseline_model), num_blocks = 100, mode = "univariate")

dbs_missense_bhr <- BHR(trait1_sumstats = dbs_sumstats_missense, annotations = list(baseline_model), num_blocks = 100, mode = "univariate")

# Convert observed scale to liability scale heritability
ac_sample_prevalence <- n_cases / (n_cases + n_controls)
population_prevalence_ac <- 0.0147
obs2lia_factor <- function(K, P) {
  X <- qnorm(K, lower.tail = FALSE)
  z <- (1 / sqrt(2 * pi)) * (exp(-(X^2) / 2))
  factor <- (K * (1 - K) * K * (1 - K)) / (P * (1 - P) * (z^2))
  return(factor)
}
ac_scalingfactor <- obs2lia_factor(population_prevalence_ac, ac_sample_prevalence)
print(ac_scalingfactor)
print(dbs_ptv_bhr$mixed_model$heritabilities[1, 5])
print(dbs_ptv_bhr$mixed_model$heritabilities[2, 5])
print(dbs_missense_bhr$mixed_model$heritabilities[1, 5])
print(dbs_missense_bhr$mixed_model$heritabilities[2, 5])

# Print heritability estimates
print(paste0("DBS PTV Burden Heritability: ",
             round(dbs_ptv_bhr$mixed_model$heritabilities[1, 5] * ac_scalingfactor, 4),
             " (", round(dbs_ptv_bhr$mixed_model$heritabilities[2, 5] * ac_scalingfactor, 4), ")"))

print(paste0("DBS Missense MPC >= 2 Burden Heritability: ",
             round(dbs_missense_bhr$mixed_model$heritabilities[1, 5] * ac_scalingfactor, 4),
             " (", round(dbs_missense_bhr$mixed_model$heritabilities[2, 5] * ac_scalingfactor, 4), ")"))
```



