---
title: "Genetic architecture of rare coding variation in mental disorders"
author: ""
date: "2024-09-06"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(data.table)
library(bhr)
```


## Introduction

Weiner et al published BHR paper in 2023 and introduced the burden heritability. We propose to apply BHR to autism and Epilepsy together with bipolar and scz. We use the variant level data from https://schema.broadinstitute.org/ in our project.


```{r}
baseline_model <- read.table("/Users/sky/Desktop/Pre_Project/Week_1/ms_baseline_oe5.txt")

bp_variantlevel_bipex <- bigreadr::fread2("/Users/sky/Desktop/Pre_Project/Week_1/BipEx_variant_results.tsv.bgz")

variant_filter = bp_variantlevel_bipex$group == "Bipolar Disorder" & #Filter to Bipolar Disorder counts
  bp_variantlevel_bipex$in_analysis == TRUE & #Use variant filter from Palmer et al, 2022
  !is.na(bp_variantlevel_bipex$gene_id) & #Remove variants with NA gene ID 
  str_detect(bp_variantlevel_bipex$locus, "^chr\\d") #Subset to autosomal variants

  
#Subset variants 
bp_variantlevel_bipex <- bp_variantlevel_bipex[variant_filter,
                                               c("gene_id",
                                                 "consequence",
                                                 "ac_case",
                                                 "ac_ctrl",
                                                 "locus",
                                                 "mpc")]  

wrangle_sumstats <- function(table,n_cases,n_controls, var_filter) {
  
  #Filter to variants of interest
  table = table[var_filter,] 
  
  #Compute sample prevalence, will be used to compute per-sd beta
  prevalence = n_cases/(n_cases+n_controls) 
  
  #Compute variant MAF in cases, will be used to compute per-sd beta
  table$AF_case = table$ac_case/(2*n_cases) 
  
  #Compute variant MAFoverall, will be used to compute per-sd beta
  table$AF = (table$ac_case + table$ac_ctrl)/(2*(n_cases + n_controls)) 
  
  #calculate per-sd betas
  table$beta = (2 * (table$AF_case - table$AF) * prevalence)/sqrt(2 * table$AF * (1 - table$AF) * prevalence * (1 - prevalence))  
  
  #calculate variant variances
  table$twopq = 2*table$AF * (1 - table$AF) 
  
  #convert betas from per-sd (i.e. sqrt(variance explained)) to per-allele (i.e. in units of phenotype) betas.
  #per-allele are the usual betas reported by an exome wide association study.
  table$beta_perallele = table$beta/sqrt(table$twopq)
  
  #aggregate into gene-level table.
  #position doesn't need to be super precise, as it is only used to order genes for jackknife
  #N = sum of case and control counts
  sumstats = data.frame(gene = table$gene_id,
                        AF = table$AF,
                        beta = table$beta_perallele,
                        gene_position = parse_number(sapply(strsplit(table$locus, split = ":"), function(x) x[[2]])),
                        chromosome =  parse_number(sapply(strsplit(table$locus, split = ":"), function(x) x[[1]])),
                        N = n_cases + n_controls,
                        phenotype_key = "BP")
  
  #we have found that in these smaller sample analyses, there are some genes with
  #large burden scores that are clearly outliers
  #we remove one such gene, TTN
  sumstats <- sumstats[sumstats$gene != "ENSG00000155657",]
  
  
  return(sumstats)
}

bp_sumstats_ptv = wrangle_sumstats(bp_variantlevel_bipex,
                                             14210, #N from Palmer et al, 2022
                                             14422, #N from Palmer et al, 2022
                                             bp_variantlevel_bipex$consequence == "ptv")
saveRDS(bp_sumstats_ptv, file = "bp_sumstats_ptv.rds")
bp_sumstats_missenseMPC2 = wrangle_sumstats(bp_variantlevel_bipex[!is.na(bp_variantlevel_bipex$mpc),],
                                             14210,
                                             14422,
                                              bp_variantlevel_bipex$consequence[!is.na(bp_variantlevel_bipex$mpc)] %in% c("damaging_missense", "other_missense") &
                                              bp_variantlevel_bipex$mpc[!is.na(bp_variantlevel_bipex$mpc)] > 2)

bp_sumstats_synonymous = wrangle_sumstats(bp_variantlevel_bipex,
                                             14210,
                                             14422,
                                             bp_variantlevel_bipex$consequence == "synonymous")


#install BHR
#devtools::install_github("ajaynadig/bhr", force = TRUE)

bp_ptv_bhr <- BHR(trait1_sumstats = bp_sumstats_ptv, 
                       annotations = list(baseline_model), #baseline model including constraint annotations
                       num_blocks = 100, #number of blocks for jackknife
                       mode = "univariate") #run in univariate mode to compute burden h2


bp_missenseMPC2_bhr <- BHR(bp_sumstats_missenseMPC2, 
                       annotations = list(baseline_model),
                       num_blocks = 100,
                       mode = "univariate")

bp_synonymous_bhr <- BHR(bp_sumstats_synonymous, 
                       annotations = list(baseline_model),
                       num_blocks = 100,
                       mode = "univariate")

#convert observed scale to liability scale h2

#calculate sample prevalence
bp_sample_prevalence = 14210/(14210 + 14422) 

#this population prevalence estimate is from Ferrari et al, 2016 Bipolar Disorders
population_prevalence_bp = 0.007 


obs2lia_factor <- function(K, P){
  X <- qnorm(K,lower.tail=FALSE)
  z <- (1/sqrt(2*pi))*(exp(-(X**2)/2))
  factor <- (K*(1-K)*K*(1-K))/(P*(1-P)*(z**2))
  return(factor)
}

bp_scalingfactor = obs2lia_factor(population_prevalence_bp,bp_sample_prevalence)
print(bp_scalingfactor)

print(paste0("BP PTV Burden Heritability: ",
             round(bp_ptv_bhr$mixed_model$heritabilities[1,5] * bp_scalingfactor, 4),
             " ",
             "(",
             round(bp_ptv_bhr$mixed_model$heritabilities[2,5] * bp_scalingfactor, 4),
             ")"))

print(paste0("BP Missense MPC > 2 Burden Heritability: ",
             round(bp_missenseMPC2_bhr$mixed_model$heritabilities[1,5] * bp_scalingfactor, 4),
             " ",
             "(",
             round(bp_missenseMPC2_bhr$mixed_model$heritabilities[2,5] * bp_scalingfactor, 4),
             ")"))

print(paste0("BP Synonymous Burden Heritability: ",
             round(bp_synonymous_bhr$mixed_model$heritabilities[1,5] * bp_scalingfactor, 4),
             " ",
             "(",
             round(bp_synonymous_bhr$mixed_model$heritabilities[2,5] * bp_scalingfactor, 4),
             ")"))


```

```{r eval=FALSE}
run_BHR <- function(table, groupname, n_case, n_ctrl, population_prevalence){

grouptable = table


sumstats_ptv = wrangle_sumstats(grouptable,
                                             n_case,
                                             n_ctrl,
                                             grouptable$consequence == "ptv")

sumstats_missenseMPC2 = wrangle_sumstats(grouptable[!is.na(grouptable$mpc),],
                                             n_case,
                                             n_ctrl,
                             grouptable$consequence[!is.na(grouptable$mpc)] %in% c("damaging_missense", "other_missense") & grouptable$mpc[!is.na(grouptable$mpc)] > 2)

# There is no synomynos data in epi25

ptv_bhr <- BHR(trait1_sumstats = sumstats_ptv, 
                       annotations = list(baseline_model), #baseline model including constraint annotations
                       num_blocks = 100, #number of blocks for jackknife
                       mode = "univariate") #run in univariate mode to compute burden h2

missenseMPC2_bhr <- BHR(sumstats_missenseMPC2, 
                       annotations = list(baseline_model),
                       num_blocks = 100,
                       mode = "univariate")

#convert observed scale to liability scale h2

#calculate sample prevalence
sample_prevalence = n_case/(n_case + n_ctrl) 

obs2lia_factor <- function(K, P){
  X <- qnorm(K,lower.tail=FALSE)
  z <- (1/sqrt(2*pi))*(exp(-(X**2)/2))
  factor <- (K*(1-K)*K*(1-K))/(P*(1-P)*(z**2))
  return(factor)
}

scalingfactor = obs2lia_factor(population_prevalence, sample_prevalence)

print(paste0(groupname, " PTV Burden Heritability: ",
             round(ptv_bhr$mixed_model$heritabilities[1,5] * scalingfactor, 4),
             " ",
             "(",
             round(ptv_bhr$mixed_model$heritabilities[2,5] * scalingfactor, 4),
             ")"))

print(paste0(groupname, " Missense MPC > 2 Burden Heritability: ",
             round(missenseMPC2_bhr$mixed_model$heritabilities[1,5] * scalingfactor, 4),
             " ",
             "(",
             round(missenseMPC2_bhr$mixed_model$heritabilities[2,5] * scalingfactor, 4),
             ")"))
}

result <- run_BHR(bp_variantlevel_bipex, "BP", 14210, 14422, 0.007)
```



## Reference

1. https://github.com/ajaynadig/bhr
2. Weiner, D.J., Nadig, A., Jagadeesh, K.A. et al. Polygenic architecture of rare coding variation across 394,783 exomes. Nature 614, 492â€“499 (2023). https://doi.org/10.1038/s41586-022-05684-z


